test/general/signature.js

```diff
@@ -340,15 +340,12 @@ describe("Signature", function() {
     const pub_key = (await openpgp.key.readArmored(pub_key_arm1)).keys[0];
     const msg = await openpgp.message.readArmored(msg_arm1);
 
-    await priv_key_gnupg_ext.subKeys[0].keyPacket.decrypt("abcd");
-    return msg.decrypt([priv_key_gnupg_ext]).then(function(msg) {
-      return msg.verify([pub_key]).then(async verified => {
-        openpgp.stream.pipe(msg.getLiteralData(), new WritableStream());
-        expect(verified).to.exist;
-        expect(verified).to.have.length(1);
-        expect(await verified[0].verified).to.be.true;
-        expect((await verified[0].signature).packets.length).to.equal(1);
-      });
+    await priv_key_gnupg_ext.decrypt("abcd");
+    return openpgp.decrypt({ message: msg, privateKeys: [priv_key_gnupg_ext], publicKeys: [pub_key] }).then(function(msg) {
+      expect(msg.signatures).to.exist;
+      expect(msg.signatures).to.have.length(1);
+      expect(msg.signatures[0].valid).to.be.true;
+      expect(msg.signatures[0].signature.packets.length).to.equal(1);
```

Comment:
Let's add a test for signing in case we ever switch the order of signing key packet selection to prioritize primary over subkeys

Reply:
Actually also, maybe we should just add filtering for decrypted packets in the functions that choose the packet to use for signing/decrypting. For example, wouldn't the current behavior be weird if I imported a key with no primary key and no signing-capable subkey and tried to sign with it?

Reply:
Adding the filtering may just make the error handling more consistent

Reply:
> Actually also, maybe we should just add filtering for decrypted packets in the functions that choose the packet to use for signing/decrypting.

We can't do that really easily because `getSigningKey` is also used for getting the public signing key, so we'd have to add a parameter for whether we want a decrypted key. I think just sorting by `isDecrypted()` is easier.

> For example, wouldn't the current behavior be weird if I imported a key with no primary key and no signing-capable subkey and tried to sign with it?

The current behavior would be throwing "Private key is not decrypted." when you try to sign with it, whereas with that change it would throw "Could not find valid signing key packet in key". I agree that in that specific case that would make sense, but it would also be a pretty weird key, and the same change would apply to the case where it's a regular old key with a perfectly valid signing key but you just forgot to decrypt it.

Reply:
`Could not find valid signing key packet in key` to me sounds like a matching error for encrypt-subkey-only type of key that @sanjanarajan mentions.

Reply:
But then `Private key is not decrypted` for this weird edge case of a key (missing primary, encrypt only subkey, trying to sign) is still miles better compared to previous behavior (which was: cannot do anything at all with any key unless primary key is present).